File: 2024/day10/day10.hs
import System.IO
import Data.Char(digitToInt)
import Data.List

find_trailheads :: [[Char]] -> [(Int, Int)]
find_trailheads grid = [(x, y)|x<-[0..length grid-1], y<-[0..length (grid!!0)-1], (grid!!x)!!y == '0']

find_next (x, y) grid = 
    let
        neighbours = [(x-1, y), (x, y+1), (x+1, y), (x, y-1)]
        numr = length grid
        numc = length (grid!!0)
        height = digitToInt ((grid!!x)!!y)
    in
        [(a, b)|(a, b)<-neighbours, ((0<=a)&&(a<numr)&&(0<=b)&&(b<numc)), (digitToInt ((grid!!a)!!b) == (height+1))]

trails pos grid = (iterate f [pos])!!9 --9 steps...
    where
        f [] = []
        f xs = nub (concat [find_next x grid|x<-xs]) --remove nub for part 2

main = do
    handle <- openFile "day10.txt" ReadMode
    contents <- hGetContents handle
    let 
        grid = lines contents
        rnum = length grid
        cnum = length (grid!!0)
        trailheads = find_trailheads grid
        scores = [length(trails pos grid)|pos<-trailheads]
    print(sum scores)
    hClose handle

File: 2024/day10/day10.py
def find_trailheads(grid):
    trailheads = []
    numr = len(grid)
    numc = len(grid[0])
    for i in range(numr):
        for j in range(numc):
            if grid[i][j]=='0':
                trailheads.append((i, j))
    return trailheads

def find_next(pos, grid):
    next_pos = []
    x, y = pos[0], pos[1]
    numr, numc = len(grid), len(grid[0])
    height = int(grid[x][y])
    neighbours = [(x-1, y), (x, y+1), (x+1, y), (x, y-1)]
    for neighbour in neighbours:
        if (0<=neighbour[0]<numr) and (0<=neighbour[1]<numc):
            if int(grid[neighbour[0]][neighbour[1]]) == height+1:
                next_pos.append(neighbour)
    return list(set(next_pos)) #remove list(set) for part 2

with open("day10.txt", "r") as f:
    grid = f.read().split("\n")
trailheads = find_trailheads(grid)
score = 0
for trailhead in trailheads:
    trails = [trailhead]
    for i in range(0, 9):
        trails = [y for x in trails for y in find_next(x, grid)]
    score+= len(trails)

print(score)